runtime/domain.c

```diff
@@ -16,79 +18,1442 @@
+
+  domain_self = di;
+  SET_Caml_state((void*)(di->tls_area));
+
+  caml_domain_set_name("BackupThread");
+
+  CAML_EVENTLOG_IS_BACKUP_THREAD();
+
+  /* TODO: how does the backup thread interact with the eventlog infra?
+   * caml_ev_tag_self_as_backup_thread(); */
+
+  msg = atomic_load_acq (&di->backup_thread_msg);
+  while (msg != BT_TERMINATE) {
+    CAMLassert (msg <= BT_TERMINATE);
+    switch (msg) {
+      case BT_IN_BLOCKING_SECTION:
+        /* Handle interrupts on behalf of the main thread:
+         *  - must hold domain_lock to handle interrupts
+         *  - need to guarantee no blocking so that backup thread
+         *    can be signalled from caml_leave_blocking_section
+         */
+        if (caml_incoming_interrupts_queued()) {
+          if (caml_plat_try_lock(&di->domain_lock)) {
+            caml_handle_incoming_interrupts();
+            caml_plat_unlock(&di->domain_lock);
+          }
+        }
+        /* Wait safely if there is nothing to do.
+         * Will be woken from caml_leave_blocking_section
+         */
+        caml_plat_lock(&s->lock);
+        msg = atomic_load_acq (&di->backup_thread_msg);
+        if (msg == BT_IN_BLOCKING_SECTION &&
+            !caml_incoming_interrupts_queued())
+          caml_plat_wait(&s->cond);
+        caml_plat_unlock(&s->lock);
+        break;
+      case BT_ENTERING_OCAML:
+        /* Main thread wants to enter OCaml
+         * Will be woken from caml_bt_exit_ocaml
+         * or domain_terminate
+         */
+        caml_plat_lock(&di->domain_lock);
+        msg = atomic_load_acq (&di->backup_thread_msg);
+        if (msg == BT_ENTERING_OCAML)
+          caml_plat_wait(&di->domain_cond);
+        caml_plat_unlock(&di->domain_lock);
+        break;
+      default:
+        cpu_relax();
+        break;
+    };
+    msg = atomic_load_acq (&di->backup_thread_msg);
+  }
+
+  /* doing terminate */
+  atomic_store_rel(&di->backup_thread_msg, BT_INIT);
+
+  return 0;
+}
+
+static void install_backup_thread (dom_internal* di)
+{
+  int err;
+#ifndef _WIN32
+  sigset_t mask, old_mask;
+#endif
+
+  if (di->backup_thread_running == 0) {
+    CAMLassert (di->backup_thread_msg == BT_INIT || /* Using fresh domain */
+            di->backup_thread_msg == BT_TERMINATE); /* Reusing domain */
+
+    while (atomic_load_acq(&di->backup_thread_msg) != BT_INIT) {
+      /* Give a chance for backup thread on this domain to terminate */
+      caml_plat_unlock (&di->domain_lock);
+      cpu_relax ();
+      caml_plat_lock (&di->domain_lock);
+    }
+
+#ifndef _WIN32
+    /* No signals on the backup thread */
+    sigfillset(&mask);
+    pthread_sigmask(SIG_BLOCK, &mask, &old_mask);
+#endif
+
+    atomic_store_rel(&di->backup_thread_msg, BT_ENTERING_OCAML);
+    err = pthread_create(&di->backup_thread, 0, backup_thread_func, (void*)di);
+
+#ifndef _WIN32
+    pthread_sigmask(SIG_SETMASK, &old_mask, NULL);
+#endif
+
+    if (err)
+      caml_failwith("failed to create domain backup thread");
```

Comment:
Note that this will always be an uncaught exception. Is there a simple way to communicate the failure to the caller of `Domain.spawn` instead?

Reply:
I think this is a good idea. Would need to work with both calls to `install_backup_thread`; one for spawn (in `domain_thread_func`) and the other when spawn is first called from the 1st domain (in `caml_domain_spawn`).